engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,609 @@
+      return retValueInterop.asBoolean(ret);
+    } catch (UnsupportedMessageException | ArityException | UnknownIdentifierException |
+             UnsupportedTypeException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "selfInterop.hasArrayElements(selfArray)",
+      "otherInterop.hasArrayElements(otherArray)"
+  }, limit = "3")
+  boolean equalsArrays(Object selfArray, Object otherArray,
+      @CachedLibrary("selfArray") InteropLibrary selfInterop,
+      @CachedLibrary("otherArray") InteropLibrary otherInterop,
+      @Cached EqualsAnyNode equalsNode
+      ) {
+    try {
+      long selfSize = selfInterop.getArraySize(selfArray);
+      if (selfSize != otherInterop.getArraySize(otherArray)) {
+        return false;
+      }
+      for (long i = 0; i < selfSize; i++) {
+        Object selfElem = selfInterop.readArrayElement(selfArray, i);
+        Object otherElem = otherInterop.readArrayElement(otherArray, i);
+        if (!equalsNode.execute(selfElem, otherElem)) {
+          return false;
+        }
+      }
+      return true;
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Behind truffle boundary, because comparing exceptions is not expected to be on fast path.
+   * @return True if the givne exceptions are the same
+   */
+  @TruffleBoundary
+  @Specialization(guards = {
+      "selfInterop.isException(selfException)",
+      "otherInterop.isException(otherException)",
+  }, limit = "3")
+  boolean equalsExceptions(Object selfException, Object otherException,
+      @CachedLibrary("selfException") InteropLibrary selfInterop,
+      @CachedLibrary("otherException") InteropLibrary otherInterop,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    try {
+      boolean sameExceptionTypes =
+          selfInterop.getExceptionType(selfException) == otherInterop.getExceptionType(otherException);
+
+      String selfMessage = null;
+      if (selfInterop.hasExceptionMessage(selfException)) {
+        selfMessage = interop.asString(
+            selfInterop.getExceptionMessage(selfException)
+        );
+      }
+      String otherMessage = null;
+      if (otherInterop.hasExceptionMessage(otherException)) {
+        otherMessage = interop.asString(
+            otherInterop.getExceptionMessage(otherException)
+        );
+      }
+      boolean sameMessages = Objects.equals(selfMessage, otherMessage);
+
+      Object selfCause =
+          selfInterop.hasExceptionCause(selfException) ? selfInterop.getExceptionCause(selfException) : null;
+      Object otherCause =
+          otherInterop.hasExceptionCause(otherException) ? otherInterop.getExceptionCause(otherException) : null;
+      boolean sameCauses = Objects.equals(selfCause, otherCause);
+
+      return sameExceptionTypes && sameMessages && sameCauses;
```

Comment:
Just wondering if it wouldn't be better to do an early `return false` for example if the types are not equal, no sense in checking messages.

Especially notable in case of checking causes - theoretically the cause can have its own cause and we can have a lot of nesting like that - then it can actually be some significant cost. It is a very rare occurrence, but it feels like it may be worth doing the early quit.

Reply:
This whole method is behind truffle boundary, and is not expected to be on the fast path, but this is still a good idea.